Add service `modbus.read_registers`

ondras12345 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Sometimes it is useful to be able to read input or holding registers without creating a new entity with polling for them. Examples include peeking at configuration registers that only need to be read once during the commissioning of a new device.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/40744
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ondras12345 committed:
Add service modbus.read_registers

janiversen commented:
Why does the PR state "The [manifest file](https://developers.home-assistant.io/docs/creating_integration_manifest/) has all fields filled out correctly.", when manifest.json is not included in the PR ?

janiversen reviewed (changes requested):
There are a number of bigger changes needed.

However there are 2 issues, which this PR does not cover:
- The raw registers are returned, so the user needs to change the word_order, depending on the device.
- With the sensors a datatype is offered, this PR demands the user to make the sometimes complicated data conversion in an automation.

To me these items needs to be addressed, to give the use a good experience.

ondras12345 commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -169,14 +180,17 @@ async def async_stop_modbus(event: Event) -> None:
-        if ATTR_UNIT in service.data:
-            slave = int(float(service.data[ATTR_UNIT]))
+        if ATTR_UNIT in data:
+            slave = int(float(data[ATTR_UNIT]))
```
The use of `int(float(...))` seemed weird to me, but that's what the existing code used.
I didn't want to cause breaking changes in the existing services, so I opted to keep the same parsing logic.

I was forced to create the `_resolve_slave` function because otherwise I kept hitting a C901 function complexity linter error, and I don't feel qualified to refactor the whole `async_modbus_setup` function.

janiversen replied:
Nobody asked you to refactor anything !

But when adding new code, it is not always good just to copy the old code "bad" code.

janiversen commented on homeassistant/components/modbus/services.yaml:
```diff
@@ -47,6 +47,39 @@ write_register:
+      selector:
+        number:
+          min: 1
+          max: 125
```
On a serial line the total packet size have is <= 256bytes.

ondras12345 replied:
I know, but apparently modbus TCP can do more. I haven't tried it though.
125 definitely will not work on modbus RTU.

janiversen replied:
Then either you need to test, if the hub used is serial or limit to what a serial line allows !!

ondras12345 commented on homeassistant/components/modbus/strings.json:
```diff
@@ -48,6 +48,32 @@
+      "description": "Reads one or more Modbus input or holding registers.",
+      "fields": {
+        "address": {
+          "name": "[%key:component::modbus::services::write_coil::fields::address::name%]",
```
That's what all other services in `strings.json` use. E.g. `write_register`:
https://github.com/home-assistant/core/blob/ea8833342da412f7611878ad61dfcf9039fb5b7e/homeassistant/components/modbus/strings.json#L29-L50
and `restart`:
https://github.com/home-assistant/core/blob/ea8833342da412f7611878ad61dfcf9039fb5b7e/homeassistant/components/modbus/strings.json#L61-L71

I just assumed that was a way to reuse strings and did the same thing. Again, I don't really know what I'm doing.

ondras12345 commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -27,18 +27,28 @@
 from homeassistant.helpers.dispatcher import async_dispatcher_send
 from homeassistant.helpers.typing import ConfigType
 from homeassistant.util.hass_dict import HassKey
+from homeassistant.util.json import JsonValueType
```
This is the only service that returns a `ServiceResponse`, that's why it is special.
The explicit cast to `JsonValueType` at line 267 is necessary to avoid a mypy warning due to `list[int]` not being a subtype of `list[JsonValueType]` due to lists being mutable.

ondras12345 commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -4,7 +4,7 @@
 import asyncio
 from collections import namedtuple
-from typing import Any
+from typing import Any, cast
```
`cast()` is used at line 267 to avoid a mypy warning. See above conversation about `JsonValueType` for more details.

ondras12345 committed:
Remove useless casts in modbus _resolve_slave

ondras12345 committed:
Separate call parameters from return values

As requested by janiversen's review.

ondras12345 committed:
Remove use of contextlib.nullcontext

As requested by janiversen's review, although I still do not
understand what is wrong with it.

ondras12345 committed:
Separate call parameters from return values

ondras12345 commented:
Sorry for the small comments. I will stop making them and instead share my remaining thoughts and questions in this single comment after pushing some new commits.

Regarding `strings.json`: those `write_coil` references are there to reference strings in this same `strings.json` file. If I changed them to `read_registers` instead, the field would reference itself. The only other alternative I see is to copy the actual strings, which does not seem very DRY.

About the max count limit: my modbus RTU device only supports reading 16 registers at a time, and simply does not respond if I request more. However, I haven't found a 16-register limit mentioned anywhere in the standard, so I think it's just a non-compliant device. The standard says up to 125 registers can be requested. I have done some back-of-the-envelope calculations and believe 125 registers might be possible with modbus RTU (contrary to what I said in one of my earlier comments) and not just TCP. It should just fit in the 256 byte limit.
I don't think it makes sense to artificially limit the number of registers that can be read just because some devices do not follow the standard. Worst case, the read fails somewhere at the transport layer, and the user receives an error / timeout. Also, I think the number in `services.yaml` needs to be a constant, so it cannot be different for RTU / TCP.

Regarding `contextlib.nullcontext`: this use is [explicitly mentioned in pytest documentation](https://docs.pytest.org/en/stable/example/parametrize.html#parametrizing-conditional-raising). Nevertheless, I have removed it in one of the new commits.

> However there are 2 issues, which this PR does not cover:
> 
>     * The raw registers are returned, so the user needs to change the word_order, depending on the device.
> 
>     * With the sensors a datatype is offered, this PR demands the user to make the sometimes complicated data conversion in an automation.
> 
> 
> To me these items needs to be addressed, to give the use a good experience.

Regarding word order, I assume you mean something like the `swap` option in the sensor config. Is that correct? Or are you concerned about architecture endianness?

In my mind, this service is more of a low-level diagnostic tool for manual use via developer tools rather than something users should regularly call from automations.
Convenience function like datatype conversion (the user would have to provide a list of datatypes when reading multiple registers) could be added later, but I would argue that if the user needs those, he should instead create sensors.

janiversen commented:
If it is a general service, it should offer data at the same level as entities.

Also thinking of it as a tool to explore a device, coils should also be supported.

The write services will 100% surely change, I have a PR prepared with the first changes.

janiversen commented:
> `modbus.read_coil` is a good idea, but I don't have the time to implement it right now, nor do I have a device I could easily test it with. Maybe later, but preferably in a separate pull request.

yes it is better as another PR.

> 
> ```yaml
> action: modbus.read_registers
> data:
>   address: 0
>   count: 8  # must match the sum of the lengths of all specified data types
>   type: holding
>   data_type: [ 'int16', 'int32', 'float64', 'float16' ]  # 8 registers
>   swap: [None, 'word', 'word_byte', None]
> ```

Just for reference this is not an API, but a structure description....but  with a couple of differences:
- one call one datatype, and count specifies how many of that datatype.
- There are quite a number of data types missing (see sensor config).
- swap: would be better as "word_order: little/big", this is the direction all configurations are heading.
- 

Would be nice if you added it to write_register as well.

> Could you please send me a link? I haven't been able to find the relevant branch in your fork.
A lot of my branches are not public, on purpose.

ondras12345 commented:
Thanks for the feedback. I will try to implement this, but realistically I don't think I will have anything ready until next weekend. I will keep this PR marked as draft until then.

> * There are quite a number of data types missing (see sensor config)

Just to make sure you understood what I meant: the yaml I posted was an *example* service call that reads 8 registers (address 0 to 7) and converts the values to the specified types. Of course, datatypes other than the ones I have used in the example would be supported, too.

I agree one datatype per call would be an acceptable compromise. It still covers my original use case (just set the type to uint16), and it does not force the user to write long data_type lists. It is less flexible than my proposal (you cannot read a block of consecutive registers with different types using one service call), but that seems like a reasonable tradeoff.

> * swap: would be better as "word_order: little/big", this is the direction all configurations are heading.

Are there any examples of this already in use, or is it just something you are planning to do in the future?

janiversen commented:
> Thanks for the feedback. I will try to implement this, but realistically I don't think I will have anything ready until next weekend. I will keep this PR marked as draft until then.
> 
Np.

> Just to make sure you understood what I meant: the yaml I posted was an _example_ service call that reads 8 registers (address 0 to 7) and converts the values to the specified types. Of course, datatypes other than the ones I have used in the example would be supported, too.

Yes and I am happy that you agree on a single datatype pr call.

> > * swap: would be better as "word_order: little/big", this is the direction all configurations are heading.
> 
> Are there any examples of this already in use, or is it just something you are planning to do in the future?

Sorry, that is still work in progress, we work on a dramatically different schema for configuration....basically allowing the user to define a device, address by address just like the typical manual.

The word_order can be used directly in the pymodbus call.
